### ISSUE TITLE
Registering a decorator for a type that is not a generic type definition but contains generic parameters should throw error?

### ISSUE THREAD

**Issue by fsateler** [simpleinjector/SimpleInjector#394]
Hi,

Condensed version: if a type is registered that is not a generic type definition, but contains a generic type parameter, Simple Injector does not complain, but then the decorator is never applied.

Long version: This is a minified version of what I was doing (this can be used in linqpad):

```csharp
void Main()
{
    var container = new Container();
    
    container.Register(typeof(IService<>), typeof(ServiceImplementation<>));
    
    var decorator = typeof(FailingServiceDecorator<>);
    var decoratedInterface = decorator.GetInterfaces().Single();
    container.RegisterDecorator(decoratedInterface, decorator);
    
    container.Verify();
    
    var instance = container.GetInstance<IService<int>>();
    instance.GetResult().Dump();
        // 0 was printed instead of an exception being thrown
}

public interface IService<T> {
    T GetResult();
}

public class ServiceImplementation<T> : IService<T> {
    public T GetResult() => default(T);
}

public class FailingServiceDecorator<T> : IService<T> {
    private readonly IService<T> Decoratee;
    public FailingServiceDecorator(IService<T> decoratee) {
        Decoratee = decoratee;
    }
    
    public T GetResult() {
        throw new InvalidOperationException();
        // we would actually use Decoratee here normally
    }
}
```

You may be wondering why such a roundabout way of registering the decorator; in my real program there are multiple decorators which are scanned from the assembly and registered on a convention basis.

Inspecting the `decoratedInterface` variable reveals that it is not a generic type definition, but it still contains generic type parameters. Calling `GetGenericTypeDefinition` on the variable makes the example work as expected.

I titled the issue as a question because I am not fully sure if this is the root cause, and because this may be by design to enable some use case I have not thought about. But I do think there is something wrong here, and Simple Injector could have helped me to find the issue. As far as I can tell, the `FailingServiceDecorator` will *never* be instantiated by Simple Injector.

**Comment by fsateler** [comment]
I just noticed `IsConstructedGenericType` is true. This might actually be the problem.

**Comment by dotnetjunkie** [comment]
Welcome to CLR/C# generic type nastiness.

What's happening is that this type:

``` c#
typeof(FailingServiceDecorator<>).GetInterface().Single()
```

Is not the same as:

``` c#
typeof(IService<>)
```

You can check this by calling:

``` c#
var t1 = typeof(FailingServiceDecorator<>).GetInterface().Single();
var t2 = typeof(IService<>);

var same = t1 == t2;
Assert.AreEqual(t1, t2);
var same = Object.ReferenceEquals(t1, t2);
```

All three statements return false. Reason for this is that the `T` in the `IService<T>` for `typeof(FailingServiceDecorator<>).GetInterface().Single()` is a _different_ `T` than that of `typeof(TService<>)`.

This is obviously weird shit, but because of this, Simple Injector just thinks that there is no match when applying the decorator; the given type constraints do not match.

Question is whether Simple Injector should handle this (this might be tricky and error prone). It's also difficult for Simple Injector to validate this and throw an exception, because technically there are types that can be matched (it will however never be a closed type that is matched).

I'll leave this issue open, and put it on the backlog. We might try to fix it in the future.

> in my real program there are multiple decorators which are scanned from the assembly and registered on a convention basis.

Be very careful with that. Decorators usually require a strict order of registration (because you want to apply them in a certain order), but when doing convention based registration you typically loose this ordering.

**Comment by fsateler** [comment]
Hi Steven, thanks for your prompt reply.

> Reason for this is that the `T` in the `IService<T>` for `typeof(FailingServiceDecorator<>).GetInterface().Single()` is a different T than that of `typeof(TService<>)`.
>
>This is obviously weird shit, but because of this, Simple Injector just thinks that there is no match when applying the decorator; the given type constraints do not match.
>
> Question is whether Simple Injector should handle this (this might be tricky and error prone). It's also difficult for Simple Injector to validate this and throw an exception, because technically there are types that can be matched (it will however never be a closed type that is matched).

Indeed. However, note the differences I highlighted:

```csharp
new [] {
	typeof(IEnumerable<>),
	typeof(IEnumerable<int>),
	typeof(IList<>).GetInterface("IEnumerable`1")
}
.Select(t => new {
	t.IsGenericTypeDefinition,
	t.IsConstructedGenericType,
	t.ContainsGenericParameters,
})
```


|IsGenericTypeDefinition|IsConstructedGenericType|ContainsGenericParameters|
|-|-|-|
|True |False |True |
|False |True| False |
|False| True |True |

Are there times where `IsConstructedGenericType` and `ContainsGenericParameters` are both true that are relevant for SimpleInjector? The only other way I could manage to generate such a type was with something as contrived as this: `typeof(IDictionary<,>).MakeGenericType(typeof(IDictionary<,>).GetGenericArguments()[0], typeof(int))`. I don't think C# language rules will let you create such a type without reflection.


>>in my real program there are multiple decorators which are scanned from the assembly and registered on a convention basis.
>
>Be very careful with that. Decorators usually require a strict order of registration (because you want to apply them in a certain order), but when doing convention based registration you typically loose this ordering.

Well, by multiple I mean multiple decorator implementations of the same basic pattern (a profiler decorator in this case), for multiple services. The position at which the decorators are registered is fully known to me, as I register only this specific convention at a point I choose.

**Comment by dotnetjunkie** [comment]
> The only other way I could manage to generate such a type was with something as contrived as this: `typeof(IDictionary<,>).MakeGenericType(typeof(IDictionary<,>).GetGenericArguments()[0], typeof(int))`. I don't think C# language rules will let you create such a type without reflection.

This is actually a completely valid use case for Simple Injector. SI is one of the few (if not only) DI Container that allows working with partly closed generic types, which what you are constructing. For Simple Injector, a partially closed type is the same as a type with type constraints and Simple Injector will apply such registration conditionally based on those condtraints.

So yes, evem though you can't express this in native C#, this is valid and SI users are actually using this feature in their applications.

**Comment by fsateler** [comment]
> So yes, evem though you can't express this in native C#, this is valid and SI users are actually using this feature in their applications.

OK, that rules out that avenue. I've looked a bit more into this and found something that might be useful. [Type.GenericTypeArguments](https://docs.microsoft.com/en-us/dotnet/core/api/system.type#System_Type_GenericTypeArguments) returns an empty array for generic type definitions, but not for constructed types. People use partially closed types. Does it make sense to use a constructed type, but that all generic type arguments have `IsGenericParameter == true` ?

In other words, SI would require that types passed to it be: 1. Generic type definitions, 2. Partially closed generic types with at least 1 concrete type bound, 3. Fully closed generic types or 4. Non-generic types.

How does that sound?

**Comment by dotnetjunkie** [comment]
In the end, the fix for this issue was rather easy. It was just a matter of adding [one line of code](https://github.com/simpleinjector/SimpleInjector/commit/b40843cf0060afd47c7800d5553da9aed36902e4).

After some checking, I found that it was actually possible to supply a `ConstructedGenericType` as service type to `RegisterDecorator`, but supplying the same type would fail with a descriptive exception when supplied to `Register(Type, Type)`. That's because `Register` actually verified whether the supplied type to not be, what Simple Injector considers, 'partially-closed'. `RegisterDecorator` simply lacked this check.

So I'm happy to announce that this check will be included in v4.1.
